File: 2021/src/day2.rs
#[aoc(day2, part1)]
fn part1(input: &str) -> usize {
    let mut x = 0;
    let mut y = 0;
    for mut it in input.lines().map(|l| l.split_ascii_whitespace()) {
        match (
            it.next().unwrap(),
            it.next().unwrap().parse::<usize>().unwrap(),
        ) {
            ("forward", n) => x += n,
            ("down", n) => y += n,
            ("up", n) => y -= n,
            (_, _) => unreachable!(),
        }
    }

    x * y
}

#[aoc(day2, part2)]
fn part2(input: &str) -> isize {
    let mut x = 0;
    let mut y = 0;
    let mut aim = 0;
    for mut it in input.lines().map(|l| l.split_ascii_whitespace()) {
        match (
            it.next().unwrap(),
            it.next().unwrap().parse::<isize>().unwrap(),
        ) {
            ("forward", n) => {
                x += n;
                y += n * aim;
            }
            ("down", n) => aim += n,
            ("up", n) => aim -= n,
            (_, _) => unreachable!(),
        }
    }

    x * y
}

File: 2021/python/day2.py
#!/usr/bin/env python3

from aoc import *


def part1():
    depth = 0
    position = 0
    def up(u): nonlocal depth; depth -= int(u)
    def down(u): nonlocal depth; depth += int(u)
    def forward(u): nonlocal position; position += int(u)
    switch(forward=forward, up=up, down=down)(puzzle.delimited())
    return depth * position


def part2():
    aim = 0
    depth = 0
    position = 0
    def up(u): nonlocal aim; aim -= int(u)
    def down(u): nonlocal aim; aim += int(u)

    def forward(u):
        nonlocal position, depth
        position += int(u)
        depth += int(u) * aim

    switch({
        'forward': forward,
        'up': up,
        'down': down,
    })(puzzle.delimited())
    return depth * position


print("Part 1:", part1())
print("Part 2:", part2())
